storm-core/src/jvm/org/apache/storm/daemon/StormCommon.java

```diff
@@ -0,0 +1,604 @@
+                Integer taskNum = Utils.getInt(conf.get(Config.TOPOLOGY_TASKS), 0);
+                if (taskNum > 0 && parallelismHintNum <= 0) {
+                    throw new InvalidTopologyException("Number of executors must be greater than 0 when number of tasks is greater than 0");
+                }
+            }
+        }
+    }
+
+    private static Set<String> getStreamOutputFields(Map<String, StreamInfo> streams) {
+        Set<String> outputFields = new HashSet<String>();
+        if (streams != null) {
+            for (StreamInfo streamInfo : streams.values()) {
+                outputFields.addAll(streamInfo.get_output_fields());
+            }
+        }
+        return outputFields;
+    }
+
+    public static void validateStructure(StormTopology topology) throws InvalidTopologyException {
+        Map<String, Object> componentMap = allComponents(topology);
+        for (Map.Entry<String, Object> entry : componentMap.entrySet()) {
+            String componentId = entry.getKey();
+            ComponentCommon common = getComponentCommon(entry.getValue());
+            if (common != null) {
+                Map<GlobalStreamId, Grouping> inputs = common.get_inputs();
+                for (Map.Entry<GlobalStreamId, Grouping> input : inputs.entrySet()) {
+                    String sourceStreamId = input.getKey().get_streamId();
+                    String sourceComponentId = input.getKey().get_componentId();
+                    if(componentMap.keySet().contains(sourceComponentId) == false) {
+                        throw new InvalidTopologyException("Component: [" + componentId + "] subscribes from non-existent component [" + sourceComponentId + "]");
+                    }
+
+                    ComponentCommon sourceComponent = getComponentCommon(componentMap.get(sourceComponentId));
+                    if (sourceComponent == null || sourceComponent.get_streams().containsKey(sourceStreamId) == false) {
+                        throw new InvalidTopologyException("Component: [" + componentId + "] subscribes from non-existent stream: " +
+                                "[" + sourceStreamId + "] of component [" + sourceComponentId + "]");
+                    }
+
+                    Grouping grouping = input.getValue();
+                    if (Thrift.groupingType(grouping) == Grouping._Fields.FIELDS) {
+                        List<String> fields = grouping.get_fields();
+                        Map<String, StreamInfo> streams = sourceComponent.get_streams();
+                        Set<String> sourceOutputFields = getStreamOutputFields(streams);
+                        if (sourceOutputFields.containsAll(fields) == false) {
+                            throw new InvalidTopologyException("Component: [" + componentId + "] subscribes from stream: [" + sourceStreamId  +"] of component " +
+                                    "[" + sourceComponentId + "] + with non-existent fields: " + fields);
```

Comment:
In the original code the error message was just for the diff-fields.  The ones that were a problem, not all of the fields.